compiler/passes/src/storage_lowering/program.rs

```diff
@@ -0,0 +1,137 @@
+            .collect::<Vec<_>>();
+
+        // After reconstructing the storage variables, `self.new_mappings` might be non-empty.
+        // Reconstruct old mappings and then append the new mappings to the final list.
+        let mut mappings =
+            input.mappings.into_iter().map(|(id, mapping)| (id, self.reconstruct_mapping(mapping))).collect::<Vec<_>>();
+        mappings.extend(self.new_mappings.clone().into_iter().collect::<Vec<_>>());
+
+        ProgramScope {
+            program_id: input.program_id,
+            consts: input
+                .consts
+                .into_iter()
+                .map(|(i, c)| match self.reconstruct_const(c) {
+                    (Statement::Const(declaration), _) => (i, declaration),
+                    _ => panic!("`reconstruct_const` can only return `Statement::Const`"),
+                })
+                .collect(),
+            structs: input.structs.into_iter().map(|(i, c)| (i, self.reconstruct_struct(c))).collect(),
+            mappings,
+            storage_variables,
+            functions: input.functions.into_iter().map(|(i, f)| (i, self.reconstruct_function(f))).collect(),
+            constructor: input.constructor.map(|c| self.reconstruct_constructor(c)),
+            span: input.span,
+        }
+    }
+
+    fn reconstruct_storage_variable(&mut self, input: StorageVariable) -> StorageVariable {
+        // For each storage variable, create one or more backing mappings that represent
+        // how the value is stored in lower-level form.
+        //
+        // Example:
+        //   storage x: field;
+        // becomes:
+        //   mapping x__: bool => field;
+        //
+        // Example (vector):
+        //   storage vec: [u32];
+        // becomes:
+        //   mapping vec__: u32 => u32;       // element mapping
+        //   mapping vec__len__: bool => u32; // length tracking
+
+        let id = || self.state.node_builder.next_id();
+        let name = input.identifier.name.to_string();
+
+        // Common base mapping name: e.g. "x__" or "vec__"
+        let mapping_name = Symbol::intern(&(name.clone() + "__"));
+
+        match &input.type_ {
+            Type::Vector(VectorType { element_type }) => {
+                // === Vector storage ===
+                //
+                // Create two mappings:
+                //   1. `<name>__`:  index → element
+                //   2. `<name>__len__`:  bool → length
+
+                // Mapping for the vector’s contents
+                self.new_mappings.insert(mapping_name, Mapping {
+                    identifier: Identifier::new(mapping_name, id()),
+                    key_type: Type::Integer(IntegerType::U32),
+                    value_type: *element_type.clone(),
+                    span: input.span,
+                    id: id(),
+                });
+
+                // Mapping for the vector’s length
+                let len_name = Symbol::intern(&(name + "__len__"));
```

Comment:
At some point, we should document our internal naming conventions and have "one source of truth". Could address with the debug information work.